packages/@aws-cdk/cfn-include/lib/cdk-include.ts

```diff
@@ -0,0 +1,221 @@
+    const ret = this.resources[logicalId];
+    if (ret) {
+      return ret;
+    }
+
+    const resourceConfig: any = this.template.Resources[logicalId];
+    const l1ClassFqn = this.cfnType2L1Class[(resourceConfig as any).Type];
+    if (l1ClassFqn) {
+      const [moduleName, ...className] = l1ClassFqn.split('.');
+      const module = require(moduleName);
+      const jsClassFromModule = module[className.join('.')];
+      // translate the properties from the template to the casing expected by CDK objects
+      const l1Props = jsClassFromModule.cfnProps2CdkProps((resourceConfig as any).Properties);
+      const l1Instance = new jsClassFromModule(this, logicalId, this.template2JsValue(l1Props));
+      if (this.preserveLogicalIds) {
+        // override the logical ID to match the original template
+        l1Instance.overrideLogicalId(logicalId);
+      }
+      this.resources[logicalId] = l1Instance;
+
+      // handle all non-property configuration
+      // (retention policies, conditions, metadata, etc.)
+      const cfnOptions: core.ICfnResourceOptions = l1Instance.cfnOptions;
+      cfnOptions.deletionPolicy = toCfnDeletionPolicy((resourceConfig as any).DeletionPolicy);
+      cfnOptions.updateReplacePolicy = toCfnDeletionPolicy((resourceConfig as any).UpdateReplacePolicy);
+      // ToDo handle:
+      // 1. Condition
+      // 2. Metadata
+      // 3. CreationPolicy
+      // 4. UpdatePolicy
+      // 5. DependsOn
+
+      return l1Instance;
+    } else {
+      throw new Error(`Unrecognized CloudFormation resource type: '${resourceConfig.Type}'`);
+    }
+  }
+
+  private template2JsValue(value: any): any {
+    if (value === undefined || value === null) {
+      return undefined;
+    }
+    if (typeof value === 'string' || typeof value === 'number') {
+      return value;
+    }
+    if (Array.isArray(value)) {
+      return value.map(el => this.template2JsValue(el));
+    }
+    if (typeof value === 'object') {
+      // An object can be either a CFN intrinsic, or an actual object
+      const cfnIntrinsic = this.serializeIfCfnIntrinsic(value);
+      if (cfnIntrinsic) {
+        return cfnIntrinsic;
+      }
+      const ret: any = {};
+      for (const [key, val] of Object.entries(value)) {
+        ret[key] = this.template2JsValue(val);
+      }
+      return ret;
+    }
+  }
+
+  private serializeIfCfnIntrinsic(object: any): any {
+    const key = looksLikeCfnIntrinsic(object);
+    switch (key) {
+      case undefined:
+        return undefined;
+      case 'Ref': {
+        const specialCaseRef = specialCaseRefs(object[key]);
+        return specialCaseRef
+          ? specialCaseRef
+          // we can't simply use object[key] here,
+          // because the logical ID in CDK might be different than the logical ID in the template
+          // (because preserveLogicalIds might be false)
+          : core.Fn.ref(this.getOrCreateResource(object[key]).logicalId);
```

Comment:
This would need to be a `CfnReference` if you want cross-stack references to work for these. That class might be private though, have a look.

Reply:
I see the need for `getOrCreateResource` now... if we did the lookup here lazily, we could be more crisp about the lifecycle of the constructs.

----

Also -- what happens if this doesn't actually refer to a logical ID that's in the same template fragment? Right now it will fail.

Just to mention some... unnamed internal tool that does merging of CFN template fragments... that people might want to use this library in combination with. If you enforce the constraint that every template fragment must be a closed world, that won't work.

Reply:
> I see the need for `getOrCreateResource` now... if we did the lookup here lazily, we could be more crisp about the lifecycle of the constructs.

I don't get what this means. We currently eagerly create all constructs - how can we be more crisp about the lifecycle of the constructs than that?

Reply:
> Also -- what happens if this doesn't actually refer to a logical ID that's in the same template fragment? Right now it will fail.
>
> Just to mention some... unnamed internal tool that does merging of CFN template fragments... that people might want to use this library in combination with. If you enforce the constraint that every template fragment must be a closed world, that won't work.

I'm fine with this initial version only working for entire templates, and failing for template fragments.

If we want to add support for template fragments later, we can.

Reply:
> This would need to be a `CfnReference` if you want cross-stack references to work for these. That class might be private though, have a look.

We actually can't implement that right now, because I gave up on threading a `resolver` through the translation that would allow you to lookup the `CfnElement` that is being targeted by the given `Ref` or `Fn::GetAtt` in this PR. But it's a great comment, and will add it to the list of tasks to complete.

I checked, and while the class `CfnReference` is module-private, now that the translation logic has been moved to `@aws-cdk/core`, we should be able to get access to it.